custom/icds/messaging/indicators.py

```diff
@@ -368,35 +366,60 @@ class LSVHNDSurveyIndicator(LSIndicator):
     def __init__(self, domain, user):
         super(LSVHNDSurveyIndicator, self).__init__(domain, user)
 
+        self.awc_ids_not_in_timeframe = get_awcs_with_old_vhnd_date(
+            domain,
+            [l.location_id for l in self.awc_locations]
+        )
+
         self.forms = get_last_form_submissions_by_user(
             domain, self.aww_user_ids, xmlns=VHND_SURVEY_XMLNS
         )
 
     def get_messages(self, language_code=None):
-        now_date = self.now.date()
-        user_ids_with_forms_in_time_frame = set()
-        for user_id, forms in self.forms.items():
-            vhnd_date = forms[0]['form'].get('vhsnd_date_past_month')
-            if vhnd_date is None:
-                continue
-            if (now_date - string_to_datetime(vhnd_date).date()).days < 37:
-                user_ids_with_forms_in_time_frame.add(user_id)
-
-        awc_ids = {
-            loc
-            for loc, user_ids in self.user_ids_by_location_id.items()
-            if user_ids.isdisjoint(user_ids_with_forms_in_time_frame)
-        }
         messages = []
 
-        if awc_ids:
-            awc_names = {self.awc_locations[awc] for awc in awc_ids}
+        if self.awc_ids_not_in_timeframe:
+            awc_names = {self.awc_locations[awc] for awc in self.awc_ids_not_in_timeframe}
             context = {'location_names': ', '.join(awc_names)}
             messages.append(self.render_template(context, language_code=language_code))
 
         return messages
 
 
+def get_awcs_with_old_vhnd_date(domain, awc_location_ids):
+    return set(awc_location_ids) - get_awws_in_vhnd_timeframe(domain)
+
+
+@icds_quickcache(timeout=12 * 60 * 60, memoize_timeout=12 * 60 * 60, session_function=None)
+def get_awws_in_vhnd_timeframe(domain):
+    # This function is called concurrently by many tasks.
+    # The CriticalSection ensures that the expensive operation is not triggered
+    #   by each task again, the compute_awws_in_vhnd_timeframe itself is cached separately,
+    #   so that other waiting tasks could lookup the value from cache.
+    with CriticalSection(['compute_awws_in_vhnd_timeframe']):
+        return compute_awws_in_vhnd_timeframe(domain)
+
+
+@icds_quickcache(timeout=60 * 60, memoize_timeout=60 * 60, session_function=None)
+def compute_awws_in_vhnd_timeframe(domain):
+    """
+    This computes awws with vhsnd_date_past_month less than 37 days.
+
+    Result is cached in local memory, so that indvidual reminder tasks
+    per AWW/LS dont hit the database each time
+    """
+    table = get_table_name(domain, 'static-vhnd_form')
+    query = """
+    SELECT DISTINCT awc_id
+    FROM "{table}"
+    WHERE vhsnd_date_past_month > '{{old}}'
+    """.format(table=table, old=datetime.today().date() - timedelta(days=37))
+
+    with connections[get_icds_ucr_citus_db_alias()].cursor() as cursor:
```

Comment:
for future safety it would be better to use the engine configured for the UCR:
```python
ds = StaticDataSourceConfiguration.by_id(StaticDataSourceConfiguration.get_doc_id(domain 'static-vhnd_form'))
django_db = connections.get_django_db_alias(ds.engine_id)
with connections[django_db].cursor() as cursor:
   ...
```

Reply:
Good point, updated.